### ISSUE TITLE
Cannot copy between a TensorFlowLite tensor with shape [1, 38, 38, 255] and a Java object with shape [1, 38, 38, 3, 85]

### ISSUE THREAD

**Issue by spaul13** [wics1224/yolov3-android-tflite#9]
I have trained a new tiny-yolo model whose input size is 608 on coco dataset.

Hence I have changed the line below
**super(assetManager, "mymodel.tflite", "coco.txt", 608);**
also changed
**mOutWidth = new int[]{38}; //previously {13, 26} for initial tiny-yolo version** 

But while detection I am getting the following error,
_**java.lang.IllegalArgumentException: Cannot copy between a TensorFlowLite tensor with shape [1, 38, 38, 255] and a Java object with shape [1, 38, 38, 3, 85].**
        at org.tensorflow.lite.Tensor.throwIfShapeIsIncompatible(Tensor.java:342)
        at org.tensorflow.lite.Tensor.throwIfDataIsIncompatible(Tensor.java:305)
        at org.tensorflow.lite.Tensor.copyTo(Tensor.java:183)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:164)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:296)
        at com.amitshekhar.tflite.Classifier.RecognizeImage(Classifier.java:289)
        at com.amitshekhar.tflite.MainActivity$DetectThread.run(MainActivity.java:84)_

Can anyone please tell me how to resolve this issue?

**Comment by avantgardnerio** [comment]
The author has trained the tiny version of the classifier on his own data, so the output tensor shape is the number of classes he trained on, not the 85 COCO default. The non-tiny version is unaffected. The solution is to download the darknet tiny weights and run yad2k.py on them and it will resolve the issue.

**Comment by zxj11838** [comment]
@spaul13 Is your problem solved? I encountered the same problem

**Comment by abdalkhalik** [comment]
> The author has trained the tiny version of the classifier on his own data, so the output tensor shape is the number of classes he trained on, not the 85 COCO default. The non-tiny version is unaffected. The solution is to download the darknet tiny weights and run yad2k.py on them and it will resolve the issue.

I tried it, however yad2k.py in this repo generated .h5 keras file. I tried importing it to the project but that didn't work

**Comment by abdalkhalik** [comment]
> > The author has trained the tiny version of the classifier on his own data, so the output tensor shape is the number of classes he trained on, not the 85 COCO default. The non-tiny version is unaffected. The solution is to download the darknet tiny weights and run yad2k.py on them and it will resolve the issue.
> 
> I tried it, however yad2k.py in this repo generated .h5 keras file. I tried importing it to the project but that didn't work

It seem to be working now, i imported yolo3_tiny.tflite, used the coco.txt and flipped to **mOutWidth = new int[]{26,13};**
and it worked, but the app cannot detect anything. I am not sure its caused by what i changed

**Comment by zxj11838** [comment]
@abdalkhalik 
My tflite model is already working, I modified the RecognizeImage () method, Modify the following code：
float[][][][][] out = new float[1][mOutWidth[i]][mOutWidth[i]][3][5 + mLabelList.size()];
to：
float[][][][] out = new float[1][mOutWidth[i]][mOutWidth[i]][3*(5 + mLabelList.size())];
And modify subsequent index operations related to "out".

**Comment by abdalkhalik** [comment]
> @abdalkhalik
> My tflite model is already working, I modified the RecognizeImage () method, Modify the following code：
> float[][][][][] out = new float[1][mOutWidth[i]][mOutWidth[i]][3][5 + mLabelList.size()];
> to：
> float[][][][] out = new float[1][mOutWidth[i]][mOutWidth[i]][3*(5 + mLabelList.size())];
> And modify subsequent index operations related to "out".

 thank you @zxj11838, can you share the modified code if its possible

**Comment by zxj11838** [comment]
@abdalkhalik    

There are very few changes, mainly modify the five-dimensional index operation into a four-dimensional index operation, you can refer to it.  As follows:

public ArrayList<Recognition> RecognizeImage(Bitmap bitmap) {
        ByteBuffer byteBuffer = convertBitmapToByteBuffer(bitmap);

        Map<Integer, Object> outputMap = new HashMap<>();
        for (int i = 0; i < mOutWidth.length; i++) {
            float[][][][] out = new float[1][mOutWidth[i]][mOutWidth[i]][3*(5 + mLabelList.size())];
            outputMap.put(i, out);
        }

        Log.d("wangmin", "mObjThresh: " + getObjThresh());

        Object[] inputArray = {byteBuffer};
        mInterpreter.runForMultipleInputsOutputs(inputArray, outputMap);

        ArrayList<Recognition> detections = new ArrayList<Recognition>();

        for (int i = 0; i < mOutWidth.length; i++) {
            int gridWidth = mOutWidth[i];
            float[][][][] out = (float[][][][])outputMap.get(i);

            Log.d("wangmin", "out[" + i + "] detect start");
            for (int y = 0; y < gridWidth; ++y) {
                for (int x = 0; x < gridWidth; ++x) {
                    for (int b = 0; b < NUM_BOXES_PER_BLOCK; ++b) {

                        final int offset =
                                (gridWidth * (NUM_BOXES_PER_BLOCK * (mLabelList.size() + 5))) * y
                                        + (NUM_BOXES_PER_BLOCK * (mLabelList.size() + 5)) * x
                                        + (mLabelList.size() + 5) * b;
                        // 得到每一个框的置信度
                        final float confidence = expit(out[0][y][x][b * (5 + mLabelList.size())+4]);
                        int detectedClass = -1;
                        float maxClass = 0;

                        final float[] classes = new float[mLabelList.size()];
                        for (int c = 0; c < mLabelList.size(); ++c) {
                            classes[c] = out[0][y][x][b * (5 + mLabelList.size()) + 5+c];
                        }
                        softmax(classes);

                        for (int c = 0; c < mLabelList.size(); ++c) {
                            if (classes[c] > maxClass) {
                                detectedClass = c;
                                maxClass = classes[c];
                            }
                        }
                        // 得到每一个框的类别置信度
                        final float confidenceInClass = maxClass * confidence;

                        // 对于大于指定阈值的框，计算框的左上右下坐标表示(相对于输入是mInputSize的表示).
                        if (confidenceInClass > getObjThresh()) {
                            final float xPos = (x + expit(out[0][y][x][b* (5 + mLabelList.size()) + 0])) * (mInputSize / gridWidth);
                            final float yPos = (y + expit(out[0][y][x][b * (5 + mLabelList.size()) + 1])) * (mInputSize / gridWidth);

                            final float w = (float) (Math.exp(out[0][y][x][b * (5 + mLabelList.size()) + 2]) * mAnchors[2 * mMasks[i][b] + 0]);
                            final float h = (float) (Math.exp(out[0][y][x][b * (5 + mLabelList.size()) + 3]) * mAnchors[2 * mMasks[i][b] + 1]);

                            Log.d("wangmin","box x:" + xPos + ", y:" + yPos + ", w:" + w + ", h:" + h);

                            final RectF rect =
                                    new RectF(
                                            Math.max(0, xPos - w / 2),
                                            Math.max(0, yPos - h / 2),
                                            Math.min(bitmap.getWidth() - 1, xPos + w / 2),
                                            Math.min(bitmap.getHeight() - 1, yPos + h / 2));
                            Log.d("wangmin", "detect " + mLabelList.get(detectedClass)
                                    + ", confidence: " + confidenceInClass
                                    + ", box: " + rect.toString());
                            // detections: ArrayList
                            detections.add(new Recognition("" + offset, mLabelList.get(detectedClass),
                                    confidenceInClass, rect, detectedClass));
                        }
                    }
                }
            }
            Log.d("wangmin", "out[" + i + "] detect end");
        }

        // 采用非极大值抑制过滤掉重复的框.
        final ArrayList<Recognition> recognitions = nms(detections);

        return recognitions;
    }
